expr/expr.go

```diff
@@ -128,63 +136,19 @@ func (eval evaluator) Eval(ctx context.Context, exp parser.Expr, from, until int
 		}
 		return results, nil
 	}
-	return EvalExpr(ctx, exp, from, until, values)
-}
-
-var _evaluator = evaluator{}
-
-func init() {
-	helper.SetEvaluator(_evaluator)
-	metadata.SetEvaluator(_evaluator)
-}
-
-// FetchAndEvalExp fetch data and evaluates expressions
-func FetchAndEvalExp(ctx context.Context, e parser.Expr, from, until int64, values map[parser.MetricRequest][]*types.MetricData) ([]*types.MetricData, merry.Error) {
-	targetValues, err := _evaluator.Fetch(ctx, []parser.Expr{e}, from, until, values)
-	if err != nil {
-		return nil, merry.Wrap(err)
-	}
-
-	res, err := _evaluator.Eval(ctx, e, from, until, targetValues)
-	if err != nil {
-		return nil, merry.Wrap(err)
-	}
-
-	for mReq := range values {
-		SortMetrics(values[mReq], mReq)
-	}
-
-	return res, nil
+	return EvalExpr(ctx, eval, exp, from, until, values)
 }
 
-func FetchAndEvalExprs(ctx context.Context, exprs []parser.Expr, from, until int64, values map[parser.MetricRequest][]*types.MetricData) ([]*types.MetricData, map[string]merry.Error) {
-	targetValues, err := _evaluator.Fetch(ctx, exprs, from, until, values)
-	if err != nil {
-		return nil, map[string]merry.Error{"*": merry.Wrap(err)}
-	}
-
-	res := make([]*types.MetricData, 0, len(exprs))
-	var errors map[string]merry.Error
-	for _, exp := range exprs {
-		evaluationResult, err := _evaluator.Eval(ctx, exp, from, until, targetValues)
-		if err != nil {
-			if errors == nil {
-				errors = make(map[string]merry.Error)
-			}
-			errors[exp.Target()] = merry.Wrap(err)
-		}
-		res = append(res, evaluationResult...)
-	}
-
-	for mReq := range values {
-		SortMetrics(values[mReq], mReq)
+// NewEvaluator create evaluator with limiter and zipper
+func NewEvaluator(limiter limiter.SimpleLimiter, zipper zipper.CarbonZipper) (*Evaluator, error) {
+	if zipper == nil {
+		return nil, ErrZipperNotInit
 	}
-
-	return res, errors
+	return &Evaluator{limiter: limiter, zipper: zipper}, nil
 }
 
-// EvalExpr is the main expression evaluator.
-func EvalExpr(ctx context.Context, e parser.Expr, from, until int64, values map[parser.MetricRequest][]*types.MetricData) ([]*types.MetricData, error) {
+// EvalExpr is the main expression Evaluator.
```

Comment:
I think the comment got renamed by accident

```suggestion
// EvalExpr is the main expression evaluator.
```